Docs: start restructure (#776)
hasura/ddn-docs: docs/quickstart.mdx

---
sidebar_position: 0
sidebar_label: Quickstart
description: "Create your first supergraph API with Hasura DDN in less than a minute."
keywords:
  - hasura ddn
  - graphql api
  - quickstart
  - getting started
  - guide
  - quickstart
hide_table_of_contents: true
is_guide: true
---

import Step from "@site/src/components/CodeStep";
import InstallTheCli from "@site/docs/_install-the-cli.mdx";
import CodeBlock from "@theme/CodeBlock";
import SimpleVideo from "@site/src/components/SimpleVideo";
import Admonition from "@theme/Admonition";

# Quickstart with Hasura DDN

<div className="flex flex-col gap-6 mb-10">
  <div className="">
    {/* If you break this onto another line it will lose its styling because the markdown rendering engine will add a <p> tag. Please leave it as a single line. Thanks! */}
    <p className="text-[1.3rem]">
      In less than <em>a minute</em> and without needing a data source connection string, you can have a supergraph API
      running locally and deployed on Hasura DDN. 🚀{" "}
    </p>
  </div>
  <div className="max-w-[750px]">

>     <SimpleVideo videoLink="https://www.youtube.com/embed/OsO6TzwFb30" />

Reviewer: Discussion: What if we just had this as a link: "Watch the video speedrun."

Author: I'm a fan of this 👍

  </div>
</div>

## Prerequisites

1. **Install the DDN CLI**

   <InstallTheCli />

2. **Install [Docker](https://docs.docker.com/engine/install/)** (needed for local development)

   The Docker based workflow helps you iterate and develop locally without deploying any changes to Hasura DDN, making
   the development experience faster and your feedback loops shorter. **You'll need Docker Compose `v2.27.1` or later.**

3. **Validate the installation**

   You can verify that the DDN CLI is installed correctly by running:

   <CodeBlock language="bash">{`ddn doctor`}</CodeBlock>

<Step
id="login"
language="bash"
code={`ddn auth login`}
heading={`Log in via the CLI`}
>

> Authenticate your CLI session, giving you access to Hasura Cloud resources. After you log in, the CLI will acknowledge
> your login and check for any updates.

Reviewer: Discussion:
```suggestion
Authenticate your CLI session to get access to Hasura Cloud resources. 
```
The "...achnowledge your login..." part isn't necessary to say and I checked and it doesn't check for updates any more. 

<img width="556" alt="Screenshot 2024-11-22 at 17 08 19" src="https://github.com/user-attachments/assets/563f7f3f-9748-4caa-b61b-244516c49ab3">

</Step>

<Step
id="supergraph-init"
language="bash"
code={'ddn supergraph init mysupergraph\ncd mysupergraph'}
heading={`Initialize a new supergraph in a new directory`}
>

> Initialize a new local project by setting up all necessary files and directories for development in the new
> `mysupergraph` directory. Once you move into this directory, you'll see your project scaffolded out for you.

Reviewer: Discussion:

```suggestion
After you switch to the new `my supergraph` directory you'll see your project files scaffolded out for you.
```

The headline is doing the main work, we don't need to repeat it in the first sentence again.

</Step>

<Step
id="connector-init"
language="bash"
code={`ddn connector init my_connector -i`}
heading={`Connect to data`}
>

> Add a new data connector to your supergraph project. From the dropdown, choose the `hasura/postgres` connector and
> connect to our sample PostgreSQL database using this URL:
> `postgresql://read_only_user:readonlyuser@35.236.11.122:5432/v3-docs-sample-app`

Reviewer: Discussion

```suggestion
Add a new data connector to your supergraph project. From the dropdown, choose the `hasura/postgres` connector, accept the suggested Connector Port and
use our sample PostgreSQL database with this connection URI:
`postgresql://read_only_user:readonlyuser@35.236.11.122:5432/v3-docs-sample-app` Leave the other values blank.
```

</Step>

<Step
id="connector-introspect"
language="bash"
code={`ddn connector introspect my_connector`}
heading={`Introspect your data source`}
>

> Introspect the data source to create Hasura metadata describing the schema and its structure. After running this, you
> should see tables present in your `app/connector/my_connector/configuration.json` file.

Reviewer: ```suggestion
Introspect the data source to create connector configuration describing the database schema. After running this, you
should see tables present in your `app/connector/my_connector/configuration.json` file.
```

</Step>

<Step
id="add-resources"
language="bash"
code={'ddn model add my_connector \'*\'\nddn command add my_connector \'*\'\nddn relationship add my_connector \'*\''}

> heading={`Add your resources`}

Reviewer: ```suggestion
heading={`Add your API resources`}
```

```suggestion
heading={`Add your metadata resources`}
```

>

> Create metadata for [models](/supergraph-modeling/models.mdx), [commands](/supergraph-modeling/commands.mdx), and
> [relationships](/supergraph-modeling/relationships.mdx) in your supergraph. These define the structure, operations, and
> connections within your supergraph and are generated as Hasura Metadata Language (HML) present in your `app/metadata`
> directory.

Reviewer: Discussion: 

```suggestion
Create metadata for [models](/supergraph-modeling/models.mdx), [commands](/supergraph-modeling/commands.mdx), and
[relationships](/supergraph-modeling/relationships.mdx) in your supergraph. These define the collections, operations, and
relationships within your supergraph API and are generated as Hasura Metadata Language (HML) present in your `app/metadata`
directory.
```

</Step>

> <Step
> id="build-local"
> language="bash"
> code={`ddn supergraph build local`}
> heading={`Build your supergraph`}
> >
> 
> Create an immutable build of your supergraph and the local assets to run the engine. The build is stored as a set of
> JSON files in `engine/build`.
> 
> </Step>

Reviewer: ```suggestion
<Step
id="build-local"
language="bash"
code={`ddn supergraph build local`}
heading={`Build your supergraph API`}
>

Create an immutable build of your supergraph API which runs on the engine. The build is stored as a set of JSON files in `engine/build`.

</Step>
```

<Step
id="run-local"
language="bash"
code={`ddn run docker-start`}

> heading={`Start your supergraph`}

Reviewer: Discussion

```suggestion
heading={`Start your supergraph API`}
```

>

> Start your engine, connector, and other services. You can see a list of all running services using `docker ps`.

Reviewer: Discussion

```suggestion
Start your engine, connector, and other services. 
```

</Step>

<Step
id="open-local-console"
language="bash"
code={`ddn console --local`}
heading={`Open the console`}
>

In a new terminal tab open to the project's directory, open the console — Hasura DDN's GUI — so you can test out your
API.

</Step>

<Step
id="local-query"
language="graphql"
code={`query UsersAndOrders{
  users {
    id
    name
    email
    orders {
      id 
      created_at
      status
    }
  }
}
`}
heading={`Query your data`}
>

Execute this query via your console. You'll see a list of all users and their orders returned!

</Step>

<Step
id="project-init"
language="bash"
code={`ddn project init`}

> heading={`Create a Hasura DDN project`}

Reviewer: Discussion: 

```suggestion
heading={`Create a Hasura DDN cloud project`}
```

Just for the first time use?

>

Provision a new project on Hasura DDN, which serves as the deployment environment for your supergraph. Once complete,
the CLI will return the project's name.

</Step>

<Step
id="build-create"
language="bash"
code={`ddn supergraph build create`}
heading={`Build and deploy your supergraph`}
>

Create a cloud build of your supergraph and its connector. When this process is complete, the CLI will return a link to
the hosted API where you can run the same query as before.

**That's it! you've created your very first, fully-functioning supergraph and deployed it to Hasura DDN 🎉**

</Step>

## Next steps

What we've laid out above is the quickest way to get started with Hasura DDN. However, you have complete, granular
control over each step in the process and can extend and customize your supergraph to fit your teams' needs. After

> deploying your first supergraph, take your next steps with Hasura DDN 👇

Reviewer: ```suggestion
deploying your first supergraph, take your next steps with Hasura DDN to connect to your own data sources 👇
```
